api/handlers/metadata.py

```diff
@@ -0,0 +1,187 @@
+    accountName,
+    containerName,
+    filepath,
+    response: Response,
+    db: Database = Depends(database.database.db),
+    metadata=Body(...),
+):
+    # Check datasource id is valid
+    datasource = db.datasources.find_one(
+        {"accountName": accountName, "containerName": containerName}
+    )
+    if not datasource:
+        response.status_code = 404
+        return "Datasource Not Found"
+
+    # Check metadata doesn't already exist
+    if db.metadata.find_one(
+        {
+            "accountName": accountName,
+            "containerName": containerName,
+            "filepath": filepath,
+        }
+    ):
+        response.status_code = 400
+        return {"error": "record already exists"}
+
+    with db.client.start_session() as session:
+        session.start_transaction()
+        try:
+            # Create the first metadata record
+            initial_version = {
+                "version_number": 0,
+                "accountName": accountName,
+                "containerName": containerName,
+                "filepath": filepath,
+                "metadata": metadata,
+            }
+            db.metadata.insert_one(initial_version)
+            db.versions.insert_one(initial_version)
+            session.commit_transaction()
+            return "Success"
+        except Exception as e:
+            session.abort_transaction()
+            response.status_code = 500
+            return "Transaction collision, try again: %s" % e
+
+
+def get_latest_version(db, accountName, containerName, filepath):
+    cursor = (
+        db.versions.find(
+            {
+                "accountName": accountName,
+                "containerName": containerName,
+                "filepath": filepath,
+            }
+        )
+        .sort("version", -1)
+        .limit(1)
+    )
+    result = list(cursor)
+    if not result:
+        return None
+    else:
+        return result[0]
+
+
+@router.put(
+    "/api/datasources/{accountName}/{containerName}/{filepath:path}/meta",
+    status_code=204,
+)
+def update_meta(
+    accountName,
+    containerName,
+    filepath,
+    response: Response,
+    db: Database = Depends(database.database.db),
+    metadata=Body(...),
+):
+    exists = db.metadata.find_one(
+        {
+            "accountName": accountName,
+            "containerName": containerName,
+            "filepath": filepath,
+        }
+    )
+    if exists is None:
+        response.status_code = 404
+        return {"error": "Item not found"}
+    else:
+        latest_version = get_latest_version(db, accountName, containerName, filepath)
+        if latest_version is None:
+            response.status_code = 500
+            return {"error": "No version found"}
+        with db.client.start_session() as session:
+            session.start_transaction()
+            try:
+                # This is going to be a race condition
+                version_number = latest_version["version_number"] + 1
```

Comment:
I assume this is a unique index? Just thinking what happens when someone tries to update at the same time, they would both end up with the same version number if not right?

Reply:
Also, don't we do the same call further up?
<img width="359" alt="image" src="https://github.com/IQEngine/IQEngine/assets/9397730/79c453e5-ec6f-4042-9690-517bcc4ee16e">

Reply:
Yeah this needs to be looked atr will chek this.